### DISCUSSION TITLE
How to wake Pi Pico / Pico W / Pico 2 from sleep using GPIO button?

### DISCUSSION THREAD

**Discussion by RYoder97** [micropython/micropython#16166] [Hardware & Peripherals]
A lot of people use sleep modes for finite amount of times, waking up after X seconds.

I would like to indefinitely make the microcontroller go to sleep until any of 4 GPIO buttons is pressed.

I can't seem to find a rudimentary example of this. I found some stuff related to other microcontroller boards (ESP32), but the code did not work for the Pico.

Do I use Pin.irq? What arguments should I be using?

Is the code different depending on the sleep state? Light vs deep? What are the differences?

**Comment by bidrohini** [comment]
In Light Sleep the CPU is paused, but RAM and most peripherals remain powered. In Deep Sleep, Most of the microcontroller's components, including RAM, are powered down. Only the RTC and wake-up pins remain active.
 https://www.theengineeringprojects.com/2021/12/esp32-low-power-modes.html
The Pin.irq method can be used to configure interrupts for GPIO buttons. This approach is effective for waking the Pico from light sleep.
from machine import Pin, lightsleep
import time


button1 = Pin(15, Pin.IN, Pin.PULL_UP)  # Replace 15 with your GPIO pin
button2 = Pin(16, Pin.IN, Pin.PULL_UP)
button3 = Pin(17, Pin.IN, Pin.PULL_UP)
button4 = Pin(18, Pin.IN, Pin.PULL_UP)


def wake_handler(pin):
    print(f"Button on GPIO {pin} pressed! Waking up...")


button1.irq(trigger=Pin.IRQ_FALLING, handler=wake_handler)
button2.irq(trigger=Pin.IRQ_FALLING, handler=wake_handler)
button3.irq(trigger=Pin.IRQ_FALLING, handler=wake_handler)
button4.irq(trigger=Pin.IRQ_FALLING, handler=wake_handler)


print("Entering light sleep. Press any button to wake up.")
lightsleep() 
print("Woke up!")

**Reply by peterhinch** [reply]
@bidrohini  The document you cite is about ESP32. @RYoder97 is asking about RP2.

**Reply by scruss** [reply]
I'm concerned that a lot of their responses are LLM-generated
